p2p/conn/connection.go

```diff
@@ -550,7 +559,24 @@ FOR_LOOP:
 		var err error
 		_n, err = cdc.UnmarshalBinaryLengthPrefixedReader(c.bufConnReader, &packet, int64(c._maxPacketMsgSize))
 		c.recvMonitor.Update(int(_n))
+
 		if err != nil {
+			// connection is closed (likely by the other side)
+			// shut down our side
+			if err == io.EOF {
+				c.stopServices()
+				c.Stop()
+				break FOR_LOOP
+			}
+
+			// stopServices was invoked and we are shutting down
+			// receiving is excpected to fail since we will close the connection
```

Comment:
So this happens when err != io.EOF ? What is the error here?

Reply:
This is to handle graceful shutdown via `stopServices()`. Go does not have a nice way to see that we have closed the socket - only when we go to read the socket and it fails do we need to see if we expected this failure or not. This logic queries the channel for the quit signal to notify us that we are expecting to exit and do not care about the error.